python.mk

```diff
@@ -37,107 +40,132 @@ ${PYTHON} ${ENV_DIR} ${PIP}:
 	${PIP} install --upgrade pip==9.0.1
 	${PIP} install wheel==0.30.0
 
-# Dummy targets onto which the targets defined in python_component below are
-# added as dependencies.
-${ENV_DIR}/.wheels-cleaned:
 	touch $@
 
-${ENV_DIR}/.wheels-built:
-	touch $@
+# This target builds all required wheelhouses and is therefore expected to be a
+# dependency of `make deb`
+.PHONY: wheelhouses
+
+# A dummy test target, to which dependencies are added for each test component
+.PHONY: test
 
-${ENV_DIR}/.wheels-installed:
+# Common coverage target. Because coverage is run over all components at once,
+# COVERAGE_SETUP_PY must be specified as a list of setup.py files that are run
+# under coverage.
+#
+# Optional parameters:
+#  - COVERAGE_SRC_DIR is the directory in which we run coverage.
+#  - COVERAGE_EXCL are files to be excluded from coverage
+#  - TEST_PYTHON_PATH is the python path to use
+#  - COMPILER_FLAGS are any additional flags to  be used.
+.PHONY: coverage
+coverage: ${COVERAGE} ${ENV_DIR}/.test-requirements
+	rm -rf htmlcov/
+	${COVERAGE} erase
+	# For each setup.py file in TEST_SETUP_PY, run under coverage
+	$(foreach setup, ${COVERAGE_SETUP_PY}, \
+		$(if ${TEST_PYTHON_PATH},PYTHONPATH=${TEST_PYTHON_PATH},) ${COMPILER_FLAGS} ${COVERAGE} run $(if ${COVERAGE_SRC_DIR},--source ${COVERAGE_SRC_DIR},) $(if ${COVERAGE_EXCL},--omit "${COVERAGE_EXCL}",) -a ${setup} test &&) true
+	${COVERAGE} combine
+	${COVERAGE} report -m --fail-under 100
+	${COVERAGE} html
+
+# Dummy test requirements target to which we add the requirements for each component,
+# so that coverage can depend on this target.
+${ENV_DIR}/.test-requirements:
 	touch $@
 
-# Common rules to build a python component
+# Common rules for a python component that includes tests
+# @param $1 - Target name
+#
+# Each target must supply:
+#   - <target>_TEST_REQUIREMENTS - A list of the requirements files that the target uses
+#   - <target>_TEST_SETUP        - The setup.py file used to run the tests
```

Comment:
Is it possible to default this to `$1_SETUP`?

Reply:
Yes, probably. I'll need to check other components, but I expect that defaulting it to `$1_SETUP` if not already set is the right thing to do.

Reply:
Hmm, actually I don't think that gains us much, because all the etcd components will have to override this, and all test-only components will have to set it anyway.